### PR TITLE
Fixed seed variable concatenation pointer.

### LINKED ISSUE #3819: mbedtls_ctr_drbg_reseed_internal issue
The concatenation of the seed variable is not being performed correctly in mbedtls_ctr_drbg_reseed_internal.  The call to f_entropy should pass a buffer pointer equaling seed + seedlen.
https://github.com/ARMmbed/mbedtls/blob/da9f92c2b2c38c7add6c088ca9b31bef6586b7b4/library/ctr_drbg.c#L386

### PR DESCRIPTION
**Pull request by stroebeljc**
## Description
This request fixes a minor issue with random number generation in CTR_DRBG (issue #3819).  The current implementation places the nonce value at the same starting location as the just retrieved seed value.  However, the seed size is still incremented, resulting in a section of zeroes the same length as the nonce at the end of the seed.

This fix just moves the nonce to where it is supposed to be in the concatenated seed.

Fixes #3819 

## Status
**READY**

## Backports needed?

NO - the feature with the problem is not present in any of the LTS branches.

## Steps to test or reproduce
The self-test was updated to include nonce concatenation and expected results.

### CODE REVIEW

**Comment by mpg** [comment]
Hi @stroebeljc and thanks for your contribution! Indeed the current behaviour of the code does no seem intentional: overwriting entropy with other entropy does not make sense, and is in direct contradiction with the documentation of `mbedtls_ctr_drbg_seed()` which clearly states that the outputs of the two calls to `f_entropy` are concatenated. Thanks for spotting that!

I'm taking the liberty of editing the PR description to add that backports are not needed, as the feature is not present in any of the LTS branches.

**Review by mpg — CHANGES_REQUESTED** [review]
The fix looks absolutely good to me, so I'm going to focus on what's not there :)

- I think this deserves a ChangeLog entry as a Bugfix, can you add one? (See ChangeLog.d/00README.md` for how to do that.)
- I'm not fully convinced that this can't be tested so I think it's worth at least discussing that point a bit more. In `tests/suites/test_suite_ctr_drbg.function` we use a "fixed" entropy callback, `mbedtls_test_entropy_func` for several purposes including tracking how much "entropy" was consumed and validating output against test vectors. I think the same approach should work here. IMO the questions are: (1) can we find test vectors from a trusted source that covers the desired length? and (2) if not, would it still be worth adding a test based on values generated after your patch, just to avoid any regression in the future?

Could you look a bit more into how this could bit tested and let me know what you think? (Also, once less people are on holiday, a second team member is going to review this as well and might have opinions or suggestions on this too.)

**Comment by stroebeljc** [comment]
I took a look at NIST DRBG test vectors to see if the CTR_CRBG algorithm is working as expected, but do not see alignment.  Whenever one of the Generate steps is executed, the results don't match up.  However, Instantiate and Reseed seem to match with the test vectors.  I'm not sure why this is happening.  Also, is there some testing that is run using NIST test vectors?  If so, where is that being performed?

File: library/ctr_drbg.c [hunk 1]
```diff
@@ -684,58 +684,111 @@ int mbedtls_ctr_drbg_update_seed_file( mbedtls_ctr_drbg_context *ctx,
...
 #if defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY)
-static const unsigned char result_pr[16] =
-    { 0x77, 0x84, 0x98, 0x63, 0xad, 0x37, 0xcc, 0xf3,
-      0x17, 0x66, 0xb2, 0xa5, 0x06, 0x8c, 0xa9, 0x4f };
-
-static const unsigned char result_nopr[16] =
-    { 0xd0, 0xf2, 0xc7, 0x7e, 0x12, 0x08, 0xf3, 0xcb,
-      0xf1, 0xbc, 0xd9, 0xff, 0x33, 0x6b, 0x7c, 0xa8 };
+static const unsigned char entropy_source_pr[] =
+    { 0x04, 0xd9, 0x49, 0xa6, 0xdc, 0xe8, 0x6e, 0xbb,
```
Comment (mpg) [inline library/ctr_drbg.c:689]: Thanks for using NIST test vectors here. Can you add a comment explaining where the data comes from?

**Review by mpg — CHANGES_REQUESTED** [review]
This is looking very good! I'm just requesting a comment in the self-test function documenting the origin of the test data.

File: library/ctr_drbg.c (hunk 1 above)
Comment (stroebeljc) [reply library/ctr_drbg.c:712]: I added a comment identifying the source of the test data.

**Review by mpg — APPROVED** [review]
Thanks for the reference! Looks all good to me now.

**Comment by mpg** [comment]
Now the next step is for another team member to review, which is unlikely to happen this week as most of the team is on leave, but hopefully some time next week.

File: ChangeLog.d/issue3819.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Bugfix
+   * Fix incorrect seed generation when a nonce is included
+     in mbedtls_ctr_drbg_reseed_internal.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/issue3819.txt:3]: This is not just an incorrect calculation: it's a security reduction. Effectively the nonce was all-bits-zero, so calling `mbedtls_ctr_drbg_set_nonce_len` did not increase the security strength of the initial seeding (and in particular the security strength was not as defined by NIST). Please make the description more precise and change the classification to “Security”.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Oops. Thanks for finding and fixing this! I'm mostly happy but I'd like some minor documentation improvements.

I'd have preferred to keep the self-tests as they were and improve the unit tests, but since you've already changed the unit tests, let's go with that. I've filed a [request for unit tests separately](https://github.com/ARMmbed/mbedtls/issues/3985). We shouldn't have things that are only validated through self-tests and not through specific unit tests, but it'll do for now.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from the changelog entry.

File: ChangeLog.d/issue3819.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
-Bugfix
-   * Fix incorrect seed generation when a nonce is included
-     in mbedtls_ctr_drbg_reseed_internal.
+Security
+   * Fix a security reduction error that causes all bits of the nonce
+     to be zero in mbedtls_ctr_drbg_reseed_internal.  This prevents
```
[the remark this answers is not in the document]
Comment (stroebeljc) [reply ChangeLog.d/issue3819.txt:3]: Thanks for the suggestion.  I will update.

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by mpg — APPROVED** [review]
LGTM

**Pull request merged**